gptel-transient.el

```diff
@@ -869,6 +913,302 @@ When LOCAL is non-nil, set the system message only in the current buffer."
                          (funcall quit-to-menu)))
         (local-set-key (kbd "C-c C-k") quit-to-menu)))))
 
+;; ** Suffix for displaying and removing context
+
+(defun gptel--context-edge-point (edge-type direction &optional inclusive)
```

Comment:
Preparing the context buffer can be simplified significantly.  Here's a recipe:

1. When preparing the context buffer, run through the newly added `gptel--context-overlays`, and insert the contents of each in a loop.
2. While inserting chunks, add an overlay on each segment with a `mouse-face` property.  Now each context chunk will be highlighted when hovering over it with the mouse.
3. To highlight it when the cursor is inside the overlay, you can use the `post-command-hook` as you do.  But you don't have to deal with text properties or text-property-change at all.
4. When an overlay is marked for deletion, you can simply change its face to the deletion face.

Reply:
This would allow for highlighting with the mouse cursor, but not with the caret point, which is what I assume your third point meant.

I would still have to deal with the text properties for navigating between contexts. The highlighting is already done with overlays, so I'm not sure what exactly I would be simplifying?

Reply:
> This would allow for highlighting with the mouse cursor, but not with the caret point, which is what I assume your third point meant.

Mouse cursor highlighting will be automatic once you add `mouse-face`.  My third point was about adapting the highlight to caret movement.

> I would still have to deal with the text properties for navigating between contexts. The highlighting is already done with overlays, so I'm not sure what exactly I would be simplifying?

My suggestion is to avoid using text-properties at all, and just use the overlays to navigate and track deletions.  You are essentially using text-properties like overlays here, and having to write a lot of bookkeeping code that overlays will automatically handle for you.  Since the buffer is not editable, it makes no difference to the user which of the two you use.

- When the user presses `n`/`p`, move point to beginning of the next/previous overlay.  The highlighting will be automatic from the function you add to `post-command-hook`.
- You don't need the function `gptel--context-edge-point` at all.  This is part of the bookkeeping that can be eliminated.
- To track the original overlays you can attach `gptel-context-overlay` as an overlay property instead of a text-property.

Reply:
Okay, I simplified it to use just overlays, but I didn't use `mouse-face`. I feel like having both a highlight for the mouse and the caret is too confusing. You would essentially have two highlights at the same time and I wasn't sure how marking would work. For now, I think just the caret navigation is good enough.

I did went with two overlays: one for highlighting and one for marking. I didn't want to merge them into one overlay that have a changing face because I wanted to achieve the effect of being able to highlight a deleted overlay, which works by having one overlay be superimposed on the other.

Reply:
> To track the original overlays you can attach gptel-context-overlay as an overlay property instead of a text-property.

To make some things clear, I have a function `gptel-context-string` which returns the full context string, & optionally delimits the content (i.e, the actual substring from the buffers, without the "In buffer x, line y" information). I then use this function to insert the context string into the buffer, then apply the overlays there.

You *could* argue that I could eliminate text properties entirely, but I don't know if that's smart. I would have to have separate functions to create the buffers & get the context string. Since text properties can be embedded into strings, they are a convenient way to implement this, I feel.

However, I agree with you that text properties were entirely unnecessary for the navigation of the context buffer.

Reply:
> To make some things clear, I have a function gptel-context-string which returns the full context string, & optionally delimits the content (i.e, the actual substring from the buffers, without the "In buffer x, line y" information). I then use this function to insert the context string into the buffer, then apply the overlays there.

This duplication of text can be avoided.  My suggestion is to store the original overlays (in the source buffers) as overlay properties in the context buffer.  This way you have a live reference, and don't need to use any text properties.

> You could argue that I could eliminate text properties entirely, but I don't know if it's worthwhile. I would have to have separate functions to create the buffers & get the context string. Since text properties can be embedded into strings, they are a convenient way to implement this, I feel.

I'll rewrite a couple of functions to show you what I mean, the code should be much simpler in the end.

Reply:
> This duplication of text can be avoided. My suggestion is to store the original overlays (in the source buffers) as overlay properties in the context buffer. This way you have a live reference, and don't need to use any text properties.

Bear in mind, I have user code which utilizes `gptel-context-string` *outside* of the transient menu.

> I'll rewrite a couple of functions to help show what I mean, the code should be much simpler in the end.

Thank you.